### ISSUE TITLE
exception in mapSync causes fatal error

### ISSUE THREAD

**Issue by nyurik** [dominictarr/event-stream#94]
The error thrown in the middle of the mapSync() cases application exception. Shouldn't mapSync handle that? Also, do I need to do `on('error')` after each stream transformation?
```lang=js
return new Promise(function (resolve, reject) {
    var s = fs.createReadStream(srcFile)
        .on('error', reject)
        .pipe(es.split())
        .on('error', reject)
        .pipe(es.mapSync(function (line) {
            ...
            throw new Error();
            ...
        }).on('error', reject)
        .pipe(fs.createWriteStream(outputFile))
        .on('error', reject)
        .on('finish', function () {
            resolve(outputFile);
        });
});
```

**Comment by dominictarr** [comment]
Unfortunately, there is no way other than catching `on('error')` for node streams to handle that.
(the appropiate way to handle that would depend entirely on what your map function actually does)
maybe put a try catch inside your map function if it's okay for the stream to continue.

**Comment by nyurik** [comment]
@dominictarr I think that since mapSync() expects a synchronous function, it should do the proper error handling - and wrap the call to the user func in a `try{ call() } catch{emit('error')}`. The user function should not handle its own errors by calling the stream's emit('error'), simply because otherwise you expect users to deal with errors in two ways - both sync and async (e.g. the stream breaks or the sync function fails)

**Comment by nyurik** [comment]
Fixed in v3.3.3

**Comment by nyurik** [comment]
Meh, probably should have added a unit test for this. @dominictarr if you have any spare cycles, i think it would be good to have a crashing unit test or two :)
